docs: Feature Flag Security and Compliance for Enterprises (#9528)
Unleash/unleash: website/docs/feature-flag-tutorials/use-cases/security-compliance.md

---
title: Feature Flag Security and Compliance for Enterprises
slug: /feature-flag-tutorials/use-cases/security-and-compliance
---

Security and compliance are important aspects of building and managing complex software in large enterprises. For software architects, engineering leaders, and technical decision-makers, every tool in your tech stack needs to pass security reviews. The weakest link in your software bill of materials, known as the SBOM, can be the one that compromises your security, so every dependency and integration must meet strict standards. Security isn't just about individual components—it’s about the entire system working together without introducing risk.

> In the modern security landscape, compliance frameworks like FedRamp, SOC2, and ISO 27001 set strict standards for proving good security posture in your software tool implementations. Feature flag management systems are no exception.

Reviewer: ```suggestion
In the modern security landscape, compliance frameworks like FedRAMP, SOC 2, and ISO 27001 set strict standards for proving good security posture in your software tool implementations. Feature flag management systems are no exception.
```

> It’s easy to think, “_Feature flags are just if statements_,” but managing them at scale is far more complex—especially when security and compliance are on the line. What happens if an unauthorized user turns a flag on or off in production, even if by accident? Will that jeopardize your SOC2 certification? A homegrown feature flag solution, often built for convenience, rarely has the robust access controls, audit logs, and data protection measures of a purpose-built system with contributions from thousands of developers and security experts. When compliance and security are requirements, not afterthoughts, the right tooling makes all the difference.

Reviewer: ```suggestion
It’s easy to think, “_Feature flags are just if statements_,” but managing them at scale is far more complex—especially when security and compliance are on the line. What happens if an unauthorized user turns a flag on or off in production, even if by accident? Will that jeopardize your SOC 2 certification? A homegrown feature flag solution, often built for convenience, rarely has the robust access controls, audit logs, and data protection measures of a purpose-built system with contributions from thousands of developers and security experts. When compliance and security are requirements, not afterthoughts, the right tooling makes all the difference.
```

There are several steps to ensuring your feature flag implementation is audit-proof. In this guide, we will walk through how Unleash Enterprise features can be combined to demonstrate good security posture and compliance for your organization. You will:

1. [Understand regulatory compliance frameworks in enterprise security audits](#introduction-to-regulatory-compliance-frameworks)
2. [Enable authentication controls with SSO and SCIM](#enable-authentication-controls-with-sso-and-scim)
3. [Configure role-based access control for administrators and developers](#configure-role-based-access-control-for-administrators-and-developers)
4. [Set up access controls for network security](#set-up-access-controls-for-network-security)
5. [Use a change management workflow for auditing](#use-a-change-management-workflow-for-auditing)
6. [Audit manual and automated events in Unleash](#audit-manual-and-automated-events-in-unleash)

>     - [Integrate SIEM for advanced audit compliance](#integrate-siem-for-advanced-audit-compliance)

Reviewer: what do you think about skipping the two bullet points here in the intro to make it a bit shorter?

Refer to our [security and compliance overview](/using-unleash/compliance/compliance-overview) for more details on how Unleash features measure up against these framework requirements.

## Enable authentication controls with SSO and SCIM

The first step to securely administering your feature flag system is to consider the authentication process. When you’re using Unleash, how should your team members access the platform UI? We have multiple ways for users to log into the platform, including traditional, password-based login. But when you’re considering the best approach to using feature flags in your development process, you need to consider how every access point to your projects, data, and development processes could pose a security risk for your organization. To demonstrate that your tools are configured with security and compliance in mind, we recommend you enable enterprise-grade authentication controls as your first line of defense.

> Your developers and other stakeholders need to securely access platforms used to build and ship software to production. At Unleash, we have a multi-layered authentication architecture that supports single sign-on (SSO) and SCIM protocols. These authentication controls are the de facto ways for you to integrate enterprise-grade security measures easily. Let’s take a closer look at each authentication option to use.

Reviewer: ```suggestion
Your developers and other stakeholders need to securely access platforms used to build and ship software to production. At Unleash, we have a multi-layered authentication architecture that supports [single sign-on](/reference/sso) (SSO) and [SCIM](/reference/scim) protocols. These authentication controls are the de facto ways for you to integrate enterprise-grade security measures easily. Let’s take a closer look at each authentication option to use.
```

How do you translate this into something verifiable and auditable for security reviews? What’s great is that every authentication event in Unleash is logged with detailed metadata including timestamp, IP address, user agent, and authentication method used, providing audit trails when you undergo security reviews. This shows you have a transparent, auditable system where every access can be traced, logged, and validated. So not only are you meeting compliance standards with authentication controls in place, but Unleash also automatically maintains a record for you that proves it.

### Use SCIM to automate user management at scale

Let’s consider another scenario. You are not setting up feature flag access for 10 users, but 100 or 1000 users. When these users are spread across various teams with access to different projects at various permission levels, how do you manage this effectively while maintaining a security-first approach?

User management at scale is difficult without automation. It’s also costly to manually track and change all user permissions over time.

> To solve this, Unleash uses [SCIM protocols (System for Cross-domain Identity Management)](https://scim.cloud/) to help you automatically provision and de-provision user accounts. When an employee joins or leaves your organization, their feature flag access can be automatically adjusted without manual changes. You won’t have to consider all the security implications of user accounts as your teams evolve. Unleash handles that for you. Take a look at this diagram of how Unleash automates user management at scale by enabling the SCIM integration.

Reviewer: We could remove this to be on the safe side, and add it back in in case we end up using a diagram here?

```suggestion
To solve this, Unleash uses [SCIM protocols (System for Cross-domain Identity Management)](https://scim.cloud/) to help you automatically provision and de-provision user accounts. When an employee joins or leaves your organization, their feature flag access can be automatically adjusted without manual changes. You won’t have to consider all the security implications of user accounts as your teams evolve. 
```

-   We leave audit trails, automatically [logging every permission change](#audit-manual-and-automated-events-in-unleash) with detailed user information and timestamps.
-   You can set up [approval guardrails](#use-a-change-management-workflow-for-auditing) for feature flag updates.
-   [Project isolation](/reference/project-collaboration-mode) ensures sensitive projects remain hidden from unauthorized users, while teams can only access projects relevant to their work, maintaining clear boundaries between different business units' feature flags.

Let’s look at how Unleash gives you complete control over user roles and permissions. At a high level, there are multiple [predefined roles](/reference/rbac#predefined-roles) in Unleash for you to get started with. The diagram below breaks down the root roles and their respective permissions. Think of them as very specific responsibilities only certain users have in Unleash that are defined by default.

> > **Note:** One of the key responsibilities of the _admin role_ is assigning users as project owners and members. _Editor roles_ can be designated as project owners to manage projects. _Viewer roles_ can be designated as project members to observe projects. If viewers are not assigned a project member role, they can only observe the actions taken in a project.

Reviewer: Maybe the note admonition (`:::note`) is a better fit here, slightly more consistent and helps break the text up a bit more

To get the most out of this functionality, we recommend you use change requests for your production environment to require explicit approval from at least one approver on your team. This is known as the [four-eyes principle](https://www.unido.org/overview-member-states-change-management-faq/what-four-eyes-principle). However, you can add up to 10 approvers to a submitted change request if multiple stakeholders are involved.

Let’s say a project owner wants to update a rollout strategy so that a new feature will increase from 50% exposure to 100% of your end user base. The change is approved by other team members, but the project owner will need to hold off applying the update in production until components from another team are ready to be released. Since the project owner’s changes are ready ahead of time, they’ll need to delay the change. In this case, we recommend [scheduling change requests](/reference/change-requests#scheduled-change-requests) to allow teams to plan and queue feature flag modifications with precise timing, which gives you more control and predictability in a release process.

> > **Note:** Scheduling change requests is different from directly updating a strategy constraint in a project because when you have a change request workflow configured, you’ll have to go through a whole new change request process just to update the constraint. Since the change request has already been approved in this case, it’s easier and faster to simply adjust the date of when the change will be applied.

Reviewer: Maybe a bit easier to read when we break up into two sentences?

```suggestion
> **Note:** Scheduling change requests is different from directly updating a strategy constraint in a project. When you have a change request workflow configured, you’ll have to go through a whole new change request process just to update the constraint. Since the change request has already been approved in this case, it’s easier and faster to simply adjust the date of when the change will be applied.
```

### Leverage access logs for broader auditing

Let’s think back to the importance of user management that we covered earlier. Developers and other stakeholders go through onboarding to use the platform. Authentication protocols and user provisioning ensure these processes are secure, unified, and automated. During this process, access logs keep track of who has access to Unleash for your organization’s audit compliance. You can export these logs to S3 buckets for long-term data storage. This is valuable if you need to preserve data for complying with legal or regulatory compliance, storing critical backups for disaster recovery, and archiving.

Auditing your feature flag system is made simple for traceability and reportability with Unleash’s event logs and access logs. We recommend leveraging these features as data sources for third-party services that make your data a valuable asset for security reviews, meeting compliance standards, and overall risk mitigation.

> Next, we’ll explore data protection measures built into Unleash and how to enhance data privacy with our tools.

Reviewer: I think this sentence now shouldn't be here, right? Or depends on if we still include a section on PII?

Author: Thanks! Just added the PII section that should resolve this.

-   Implement SSO and SCIM integration: leverage identity provider integration through OpenID Connect or SAML 2.0 to centralize authentication and automate user management at scale, ensuring secure and efficient onboarding and offboarding processes.

-   Use role-based access control: create custom root and project roles that follow the principle of least privilege, ensuring that team members can only access and modify feature flags relevant to their specific responsibilities.

-   Configure network security controls: restrict API access with CORS policies and IP allow lists to create multiple layers of defense around your feature flag system, limiting potential attack vectors to your production environments.

-   Establish change management workflows: implement the four-eyes principle for all production flag changes through change requests to ensure that critical modifications are properly reviewed and approved before implementation.

> -   Leverage audit logging: maintain comprehensive event logs and integrate with your SIEM systems to create a chain of evidence for all feature flag activities, which supports your compliance with SOC2, FedRAMP, and ISO 27001 requirements.

Reviewer: ```suggestion
-   Leverage audit logging: maintain comprehensive event logs and integrate with your SIEM systems to create a chain of evidence for all feature flag activities, which supports your compliance with SOC 2, FedRAMP, and ISO 27001 requirements.
```

When you engage with Unleash, you will get access to our Trust Center, which houses a complete list of annual SOC 2 and other reports, security policies, and controls, as well as a list of data sub-processors that are regularly reviewed by our compliance team and engineering leadership. These measures demonstrate Unleash’s commitment to maintaining a robust security posture, which supports our ability to pass the most rigorous enterprise security audits.

## Best practices and recommendations

Throughout this tutorial, we've explored how to use Unleash Enterprise as a secure and compliant feature flag system. You now understand how Unleash's features work together to pass security audits while enabling your teams to build and ship safely.

By using Unleash with security and compliance in mind, you've learned how to transform what could be a potential security liability into an asset that strengthens your overall security posture. Here are key recommendations to make sure your feature flag implementation meets enterprise security and compliance standards:

> -   Implement SSO and SCIM integration: leverage identity provider integration through OpenID Connect or SAML 2.0 to centralize authentication and automate user management at scale, ensuring secure and efficient onboarding and offboarding processes.

Author: Thoughts on making these bullet points stand out more? @melindafekete 
I'm tempted to embolden the words like **Implement SSO and SCIM integration**

Reviewer: Bold sounds good, thanks!

Author: Updated this 👍

---
title: Feature Flag Security and Compliance for Enterprises
slug: /feature-flag-tutorials/use-cases/security-and-compliance
---

Security and compliance are important aspects of building and managing complex software in large enterprises. For software architects, engineering leaders, and technical decision-makers, every tool in your tech stack needs to pass security reviews. The weakest link in your software bill of materials, known as the SBOM, can be the one that compromises your security, so every dependency and integration must meet strict standards. Security isn't just about individual components—it’s about the entire system working together without introducing risk.

In the modern security landscape, compliance frameworks like FedRAMP, SOC 2, and ISO 27001 set strict standards for proving good security posture in your software tool implementations. Feature flag management systems are no exception.

> It’s easy to think, “_Feature flags are just if statements_,” but managing them at scale is far more complex—especially when security and compliance are on the line. What happens if an unauthorized user turns a flag on or off in production, even if by accident? Will that jeopardize your SOC 2 certification? A homegrown feature flag solution, often built for convenience, rarely has the robust access controls, audit logs, and data protection measures of a purpose-built system with contributions from thousands of developers and security experts. When compliance and security are requirements, not afterthoughts, the right tooling makes all the difference.

Reviewer 2: It is a real thing.  In SOC2, you define your security controls and auditors check to see if you are doing what you say you will do.  This is the basis of the audit.  It is not about doing anything in particular, but more about doing what you commit to do.  So if you say that only certain people can access production (e.g. Senior Devs and SRE team), but then someone not on that list/team/etc uses feature flags to manipulate production (e.g. a Product Manager), you are at risk of falling your audit.  This is definitely not just scary-talk.

Refer to our [security and compliance overview](/using-unleash/compliance/compliance-overview) for more details on how Unleash features measure up against these framework requirements.

## Enable authentication controls with SSO and SCIM

The first step to securely administering your feature flag system is to consider the authentication process. When you’re using Unleash, how should your team members access the platform UI? We have multiple ways for users to log into the platform, including traditional, password-based login. But when you’re considering the best approach to using feature flags in your development process, you need to consider how every access point to your projects, data, and development processes could pose a security risk for your organization. To demonstrate that your tools are configured with security and compliance in mind, we recommend you enable enterprise-grade authentication controls as your first line of defense.

> Your developers and other stakeholders need to securely access platforms used to build and ship software to production. At Unleash, we have a multi-layered authentication architecture that supports [single sign-on](/reference/sso) (SSO) and [SCIM](/reference/scim) protocols. These authentication controls are the de facto ways for you to integrate enterprise-grade security measures easily. Let’s take a closer look at each authentication option to use.

Author: I rephrased so it's "industry standard authentication protocols", recognizing this is common, but that we do in fact support it.

> Your developers and other stakeholders need to securely access platforms used to build and ship software to production. At Unleash, we have a multi-layered authentication architecture that supports [single sign-on](/reference/sso) (SSO) and [SCIM](/reference/scim) protocols. These authentication controls are the de facto ways for you to integrate enterprise-grade security measures easily. Let’s take a closer look at each authentication option to use.

Reviewer: I think it would be much clearer to just say: Many SSO solutions also include two-factor authentication (2FA). I'll make a commit to add this

### Use SSO authentication for feature flags

To use single sign-on in Unleash, your users can authenticate themselves through OpenID Connect (OIDC) or SAML 2.0 protocols.

> We have integration options with enterprise identity providers like Okta, Microsoft Entra ID, and Keycloak. Connect these identity providers through either OpenID Connect or SAML 2.0 protocol. For a step-by-step guide, read our [how-to guide for single sign-on](/how-to/sso).

Reviewer 3: Dunno if this changes how you want to frame this but we support anything and everything that either has SAML or OIDC, it's just that we have comprehensive docs for the options you list

Author: Thanks for mentioning this! To me, the docs implied these were the only identity providers we supported. So I will certainly update this to make it more clear what the options are / could be.

Reviewer: Ooh I didn't know this 🤔 I'll send you an updated diagram

> Another major benefit to using SSO is that when you connect Unleash to your identity provider, you will be able to synchronize your user groups out of the box. That way, your group configuration reflects organizational-level changes automatically. Follow our step-by-step guide to set up [user group syncing](/how-to/how-to-set-up-group-sso-sync). With SSO integration, groups defined in your identity provider are directly mapped to Unleash access groups. This means:

Author: @sighphyre Okay, I'll rephrase it. I can inform the reader that group syncing _exists_, but point to SCIM which is explained in the next section as the recommended way for larger scale user management in unleash.

> Another major benefit to using SSO is that when you connect Unleash to your identity provider, you will be able to synchronize your user groups out of the box. That way, your group configuration reflects organizational-level changes automatically. Follow our step-by-step guide to set up [user group syncing](/how-to/how-to-set-up-group-sso-sync). With SSO integration, groups defined in your identity provider are directly mapped to Unleash access groups. This means:

Reviewer 3: Ye! Sounds great!

Group syncing is a viable choice if you're using a provider that doesn't implement SCIM but I think we should just not have that conversation in this doc

### Use SCIM to automate user management at scale

Let’s consider another scenario. You are not setting up feature flag access for 10 users, but 100 or 1000 users. When these users are spread across various teams with access to different projects at various permission levels, how do you manage this effectively while maintaining a security-first approach?

User management at scale is difficult without automation. It’s also costly to manually track and change all user permissions over time.

To solve this, Unleash uses [SCIM protocols (System for Cross-domain Identity Management)](https://scim.cloud/) to help you automatically provision and de-provision user accounts. When an employee joins or leaves your organization, their feature flag access can be automatically adjusted without manual changes. You won’t have to consider all the security implications of user accounts as your teams evolve.

By enabling SCIM in Unleash, you can:

-   [Provision and de-provision users](/reference/scim) (team members) as they are joining or leaving your organization.
-   Automatically create and delete user groups.

> -   Sync group membership in real-time.

Reviewer 3: SCIM isn't real time, it's up to the provider but generally around a 40 minute lag

```suggestion
-   Sync group membership.
```

-   Ensure consistent access across multiple platforms.

To unlock these benefits, set up [SCIM for automatic provisioning using our how-to guides](/how-to/provisioning).

## Configure role-based access control for administrators and developers

Now that you understand how Unleash handles authentication securely, let’s explore what needs to happen once your teams have access to the platform. When enterprise auditors evaluate software tools and development practices, they focus intensely on access control mechanisms.

> Consider this scenario: a Fortune 500 company is undergoing its annual security audit, and the auditors are specifically examining how feature flags are managed across their software development lifecycle. Without role-based access controls (RBAC), this company would face significant compliance risks and potential security vulnerabilities.

Author: Love it! Sounds good

[Role-based access control](/reference/rbac) allows for fine-grained permission management so team members can only access and modify feature flags relevant to their specific roles and responsibilities. During an audit, you'll most likely face questions about your access control systems and security measures. Here’s what auditors will want to see:

-   Only authorized personnel can modify feature flags in production.
-   All the access control changes are tracked and documented.
-   An approval chain is in place to change critical features.
-   Sensitive projects remain visible only to appropriate teams and/or project owners.

> Unleash is built with many mechanisms in place to handle all of these scenarios. Our systems enforce strict separation of duties, for example, where developers can create and test flags in development environments, while only senior engineers can promote changes to staging.

Reviewer 4: ```suggestion
Unleash is built with many mechanisms in place to handle all of these scenarios. Our systems enforce strict separation of duties, for example, where developers can create and test flags in development environments, while only senior engineers can promote changes to staging and production.
```

-   We leave audit trails, automatically [logging every permission change](#audit-manual-and-automated-events-in-unleash) with detailed user information and timestamps.
-   You can set up [approval guardrails](#use-a-change-management-workflow-for-auditing) for feature flag updates.
-   [Project isolation](/reference/project-collaboration-mode) ensures sensitive projects remain hidden from unauthorized users, while teams can only access projects relevant to their work, maintaining clear boundaries between different business units' feature flags.

Let’s look at how Unleash gives you complete control over user roles and permissions. At a high level, there are multiple [predefined roles](/reference/rbac#predefined-roles) in Unleash for you to get started with. The diagram below breaks down the root roles and their respective permissions. Think of them as very specific responsibilities only certain users have in Unleash that are defined by default.

:::note

> One of the key responsibilities of the _admin role_ is assigning users as project owners and members.

Reviewer 4: Although this can be controlled with Terraform, making it easier for admins

Reviewer 3: +1, this ties in the automation and compliance story nicely

> _Editor roles_ can be designated as project owners to manage projects.

Reviewer 4: This is not entirely true. You can be a Viewer in the system and still be a project owner or have more permissions inside a system. Usually, following the least privilege principle, the ideal would be to make all users "Viewers" and only give specific permissions. 

In terms of security recommendations we should advice to use Viewer roles or a custom root role with minimal permissions

Next, we’ll explore change management as an important tool to further enhance access control mechanisms and make your workflows auditable.

## Use a change management workflow for auditing

Engineers typically go through a development workflow where they write code, submit their code changes for review, get explicit approval from other engineers, and merge their approved changes into the main branch. This occurs in a version control system like Git/Github. In the context of feature flag management, this approach is similar to how you can submit [change requests](/reference/change-requests) in Unleash to make updates to resources in your projects across your environments.

Think of change requests as _more_ than just a tool to sign off on what’s happening in your feature flag system. It’s a way to ensure multiple layers of security for your teams.

> For large organizations in highly regulated industries, it’s important to keep track of changes in your feature flag system. When your team members are using Unleash, consider that while some users have permission to modify flags, you can still put guardrails in place to approve or deny the changes.

Reviewer 4: I think the part of auditing is covered in another section, this change would make it closer to change request and 4 eyes approval. 
```suggestion
For large organizations in highly regulated industries, it’s important to have a second pair of eyes validating a change. When your team members are using Unleash, consider that while some users have permission to modify flags, you can still put guardrails in place to approve or deny the changes.
```

Imagine a developer working at a large banking platform who wants to modify a feature flag controlling a new authentication method. Instead of making an immediate change, the [change request](/reference/change-requests) workflow could require:

1. Initial proposal submission
2. Security team review
3. Compliance officer approval
4. Final sign-off from technical leadership

With Unleash, you can create a change request workflow to reflect these exact requirements.
And from a security perspective, you'll always want an auditable trail of changes that occur in your feature flag system. We record all changes like these in the event logs. These are the configuration management practices required by compliance frameworks like SOC 2, FedRAMP, and ISO 27001. We’ll explore [event logs](#audit-manual-and-automated-events-in-unleash) more in a later section, but it’s important to note throughout this tutorial since it’s relevant to many features in Unleash that align with security and compliance standards.

> > **Note:** Anyone with project access can create a change request; however, only users with specific permissions can approve, apply, or skip them. None of the [predefined roles](/reference/rbac#predefined-roles) have any change request permissions, so you must create [custom project roles](/reference/rbac#predefined-roles).

Reviewer: @sighphyre just to confirm, do you mean that what we currently have in the note _is_ or _is not_ the default behavior? As far as I understand the current text is correct in the case of the default (open) collaboration mode.

> > **Note:** Anyone with project access can create a change request; however, only users with specific permissions can approve, apply, or skip them. None of the [predefined roles](/reference/rbac#predefined-roles) have any change request permissions, so you must create [custom project roles](/reference/rbac#predefined-roles).

Author: This note was taken from the existing Change requests doc. https://docs.getunleash.io/reference/change-requests#change-request-permissions

> > **Note:** Anyone with project access can create a change request; however, only users with specific permissions can approve, apply, or skip them. None of the [predefined roles](/reference/rbac#predefined-roles) have any change request permissions, so you must create [custom project roles](/reference/rbac#predefined-roles).

Reviewer 3: Yeah, I meant it _is not_ but now that I'm reading that again, I think it depends on what one understands by "project access". I interpret that as "project membership" but that might not be what we're talking about here

> > **Note:** Anyone with project access can create a change request; however, only users with specific permissions can approve, apply, or skip them. None of the [predefined roles](/reference/rbac#predefined-roles) have any change request permissions, so you must create [custom project roles](/reference/rbac#predefined-roles).

Reviewer: I'll work on this @nnennandukwe

> > **Note:** Anyone with project access can create a change request; however, only users with specific permissions can approve, apply, or skip them. None of the [predefined roles](/reference/rbac#predefined-roles) have any change request permissions, so you must create [custom project roles](/reference/rbac#predefined-roles).

Reviewer: applied in a5f4b4a48cc6b1146934f91d2e1198240aeafdca

![This GIF shows how to quickly make changes to your feature flag, request the change, approve it, and apply the changes.](/img/use-case-user-mgmt-cr.gif)

> To get the most out of this functionality, we recommend you use change requests for your production environment to require explicit approval from at least one approver on your team. This is known as the [four-eyes principle](https://www.unido.org/overview-member-states-change-management-faq/what-four-eyes-principle). However, you can add up to 10 approvers to a submitted change request if multiple stakeholders are involved.

Reviewer 4: I feel this part can be maybe merged with the introduction paragraphs. But it's fine, maybe redundancy and repetition has benefits here.

Author: I like that idea. 👍

> Let’s say a project owner wants to update a rollout strategy so that a new feature will increase from 50% exposure to 100% of your end user base. The change is approved by other team members, but the project owner will need to hold off applying the update in production until components from another team are ready to be released. Since the project owner’s changes are ready ahead of time, they’ll need to delay the change. In this case, we recommend [scheduling change requests](/reference/change-requests#scheduled-change-requests) to allow teams to plan and queue feature flag modifications with precise timing, which gives you more control and predictability in a release process.

Reviewer 4: Line 139 also has a similar example, maybe not so detailed...

Reviewer: Thanks! I think that example isn't about scheduling. we'll update the phrasing to distinguish them better.

Protecting data is critical within any software tool. It’s a security best practice that is ideally implemented across your tech stack where user data is handled. While completely safeguarding your systems from attacks is not foolproof, it’s better to err on the side of risk prevention and mitigation. Unleash is architecturally designed to protect organizations and their end users by keeping data private and limiting the scope of data that is evaluated and stored. Think about the types of sensitive data that could be exposed if you use a feature flag system that isn’t designed to protect data. When feature flags are evaluated, here are examples of what could be at risk:

-   API keys
-   User IDs
-   Email addresses
-   Flag data

> -   Flag configurations

Reviewer 4: Isn't flag data and flag configurations the same?

-   User geographical locations

> This is a potential attack surface area that you wouldn’t want to be the root of data breaches or unauthorized access in your system. That’s why Unleash was built in a way where integrating feature flags into your services won’t be a weak link in your organization’s ability to ship software safely. When using our [Frontend API](/reference/api/unleash/frontend-api), Unleash only stores these details in its short-term runtime cache. This is practical for development practices and applications that don’t have a lot of traffic, like internal dashboards.

Reviewer 4: ```suggestion
This is a potential attack surface area that you wouldn’t want to be the root of data breaches or unauthorized access in your system. That’s why Unleash was built in a way where integrating feature flags into your services won’t be a weak link in your organization’s ability to ship software safely. When using our [Frontend API](/reference/api/unleash/frontend-api), Unleash only uses user provided data for feature evaluation and never stores that information.
```

I would also remove: 

> This is practical for development practices and applications that don’t have a lot of traffic, like internal dashboards.

Because frontend API is mostly for frontend applications. They can tolerate a ton of traffic, which is when we recommend using edge. It's a matter of how much traffic Unleash can handle and how much it can scale, not about the purpose of the API. When Unleash can't handle much traffic we have edge, but both use the same Frontend API

Reviewer 3: > When Unleash can't handle much traffic we have edge, but both use the same Frontend API

I think it's fair to talk about the Frontend API being a separate thing from Edge, since that's how we refer to it in the documentation. It holds the same contract but "Frontend API" is a name that refers to that specific API in Unleash itself

-   Only authorized personnel can modify feature flags in production.
-   All the access control changes are tracked and documented.
-   An approval chain is in place to change critical features.
-   Sensitive projects remain visible only to appropriate teams and/or project owners.

Unleash is built with many mechanisms in place to handle all of these scenarios. Our systems enforce strict separation of duties, for example, where developers can create and test flags in development environments, while only senior engineers can promote changes to staging.

-   We leave audit trails, automatically [logging every permission change](#audit-manual-and-automated-events-in-unleash) with detailed user information and timestamps.
-   You can set up [approval guardrails](#use-a-change-management-workflow-for-auditing) for feature flag updates.
-   [Project isolation](/reference/project-collaboration-mode) ensures sensitive projects remain hidden from unauthorized users, while teams can only access projects relevant to their work, maintaining clear boundaries between different business units' feature flags.

> Let’s look at how Unleash gives you complete control over user roles and permissions. At a high level, there are multiple [predefined roles](/reference/rbac#predefined-roles) in Unleash for you to get started with. Root roles are meant to control permissions to top-level resources, spanning across all projects. Project roles, on the other hand, control permissions for a project, the feature flags, and individual configurations per environment. The diagram below breaks down the root roles, project roles, and their respective permissions. Think of them as very specific responsibilities only certain users have in Unleash that are defined by default.

Reviewer: ```suggestion
Let’s look at how Unleash gives you complete control over user roles and permissions. At a high level, there are multiple [predefined roles](/reference/rbac#predefined-roles) in Unleash for you to get started with. Root roles are meant to control permissions to top-level resources, spanning across all projects. Project roles, on the other hand, control permissions for a project, the feature flags, and individual configurations per environment.

The following diagram breaks down the root roles, project roles, and example permissions that the roles grant. 
```

I don't really get the last sentence, can we skip or rephrase?

![The diagram showing the relationship between root roles and project roles in Unleash.](/img/root-and-project-roles.png)

> :::note

Reviewer: Actually now that I see the image and the admonition box together on the live preview, I think it looks a bit strange visually. Maybe it would be better as just regular text? sorry to keep going back and forth on this

Reviewer 4: Agree, the image is not very clear. The thing is that editor can do almost everything that a project role can do. Not sure is something enterprise customers will need. I believe here on best practices and recommendations for big enterprises we should advice to make all users viewers and then have specific permissions on a per-project level, following the least privilege principle.

Author: I can remove the diagram and update the "note" to be regular text format. And add in the best practices mentioned (from Gaston).

Reviewer: We should keep the diagram, I think. I can make the changes to the text formatting

Reviewer: applied in a5f4b4a48cc6b1146934f91d2e1198240aeafdca

Securing your network layer is non-negotiable for building and using software systems. It’s foundational knowledge and good practice for any organization with engineering teams deploying software. Now that you understand how to set up and manage users at scale in Unleash, you can begin configuring Unleash in your applications and services [using our SDKs](/reference/sdks#official-sdks). Using the admin UI is an easy, direct way to make changes to your feature flags, projects, rollout strategies, and more. But when you’re making calls to our frontend API from your services, you can update cross-origin resource sharing (CORS) settings to control application access to Unleash. You can also set up IP allow lists to restrict access to Unleash.

### Set up CORS policies for Frontend API security

It was noted in [the previous section](#configure-role-based-access-control-for-administrators-and-developers) that access controls are about security measures in place at every entry point to your systems. This means they go beyond just user management. Extend these controls to [Frontend APIs](/reference/api/unleash/frontend-api) to prevent security risks across your stack.

> From an architectural perspective, we implement this by allowing Unleash users to define cross-origin resource sharing (CORS) policies to prevent unwarranted access to your feature flag system. From a compliance perspective, this demonstrates that your organization leverages a feature flag system that takes a multi-faceted approach to security.

Reviewer: Comment from Ivar:

in reality: This is a requirement from modern browser. They will not connect to another domain if it is not explicitly allow cross-origin access from the provided domain

Suggestion to change to something like:
Modern browsers enforce cross-origin resource sharing (CORS) restrictions, preventing connections to other domains unless explicitly allowed. In Unleash, you can define cross-origin resource sharing (CORS) policies to prevent unwarranted access to your feature flag system. This not only ensures compliance with browser security requirements but also reinforces a multi-layered security approach within your organization.

Author: Updated, thanks!

### Leverage access logs for broader auditing

Let’s think back to the importance of user management that we covered earlier. Developers and other stakeholders go through onboarding to use the platform. Authentication protocols and user provisioning ensure these processes are secure, unified, and automated. During this process, access logs keep track of who has access to Unleash for your organization’s audit compliance. You can export these logs to S3 buckets for long-term data storage. This is valuable if you need to preserve data for complying with legal or regulatory compliance, storing critical backups for disaster recovery, and archiving.

> Auditing your feature flag system is made simple for traceability and reportability with Unleash’s event logs and access logs. We recommend leveraging these features as data sources for third-party services that make your data a valuable asset for security reviews, meeting compliance standards, and overall risk mitigation.

Reviewer: In the Google doc there was a comment from Michael about highlighting that access logs can also tell you about API interactions as well. Maybe instead of saying: "access logs keep track of who has access to Unleash for your organization’s audit compliance" we can say something like "access logs keep track of what users and systems accessed Unleash and what actions they performed."

Author: Updated to:

"Let’s think back to the importance of user management that we covered earlier. Developers and other stakeholders go through onboarding to use the platform. Authentication protocols and user provisioning ensure these processes are secure, unified, and automated. During this process, access logs keep track of what users and systems accessed Unleash and what actions they performed, including [Unleash API interactions](/reference/api/unleash) from your services and applications."
